src/OSPSuite.Core/Domain/Services/SimModelBatch.cs

```diff
@@ -13,66 +13,108 @@ public class SimModelBatch : SimModelManagerBase, ISimModelBatch
-         setVariableParameters(simulation);
+         setVariableParameters(simulation, variableParameterPaths);
+         setVariableMolecules(simulation, variableMoleculePaths);
          FinalizeSimulation(simulation);
          return simulation;
       }
 
-      private void setVariableParameters(Simulation simulation) => SetVariableParameters(simulation, _variableParameterPaths, _calculateSensitivities);
+      private void setVariableParameters(Simulation simulation, IReadOnlyList<string> variableParameterPaths)
+      {
+         _allVariableParameters = SetVariableParameters(simulation, variableParameterPaths, _calculateSensitivities);
+         VariableParameterPaths = _allVariableParameters.Select(x => x.Path).ToList();
+      }
+
+      private void setVariableMolecules(Simulation simulation, IReadOnlyList<string> variableMoleculePaths)
+      {
+         _allVariableMolecules = SetVariableMolecules(simulation, variableMoleculePaths);
+         VariableMoleculePaths = _allVariableMolecules.Select(x => x.Path).ToList();
+      }
 
       public SimulationRunResults RunSimulation()
       {
          try
          {
-            updateParameterValues();
+            updateSimulationValues();
             _simModelSimulation.RunSimulation();
             return new SimulationRunResults(success: true, warnings: WarningsFrom(_simModelSimulation), results: getResults());
          }
          catch (Exception e)
          {
-            return new SimulationRunResults(success: false, warnings: WarningsFrom(_simModelSimulation), results: new DataRepository(), error: e.FullMessage());
+            return new SimulationRunResults(success: false, warnings: WarningsFrom(_simModelSimulation), results: new DataRepository(),
+               error: e.FullMessage());
          }
          finally
          {
             _parameterValueCache.Clear();
+            _initialValueCache.Clear();
          }
       }
 
-      public void UpdateParameterValue(string path, double value)
+      public void UpdateParameterValue(string parameterPath, double value) => _parameterValueCache[parameterPath] = value;
+
+      public void UpdateParameterValues(IReadOnlyList<double> parameterValues)
       {
-         _parameterValueCache[path] = value;
+         updateValues(parameterValues, VariableParameterPaths, _parameterValueCache, "parameter values");
+      }
+
+      public void UpdateInitialValue(string moleculePath, double value) => _initialValueCache[moleculePath] = value;
+
+      public void UpdateInitialValues(IReadOnlyList<double> initialValues)
+      {
+         updateValues(initialValues, VariableMoleculePaths, _initialValueCache, "initial values");
+      }
+
+      private void updateValues(IReadOnlyList<double> values, IReadOnlyList<string> variablePaths, ICache<string, double> valuesCache,
+         string updatedValues)
+      {
+         var count = values?.Count ?? 0;
+         if (variablePaths.Count != count)
```

Comment:
Here I check explicitely that we are passing the same number of values (either parameter values or initial amount)